File: Roads-and-Libraries.cs
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;


public class Graph
{
    public bool Directed { get; set; }
    public long NumVertices { get; set; }
    public List<List<long>> AdjacencyList { get; set; }

    public Graph(long numVertices, bool directed)
    {
        NumVertices = numVertices;
        Directed = directed;
        AdjacencyList = new List<List<long>>((int)numVertices);
        for (long i = 0; i < numVertices; i++)
        {
            AdjacencyList.Add(new List<long>());
        }
    }

    public void AddEdge(int vertex1, int vertex2)
    {
        AdjacencyList[vertex1].Add(vertex2);
        if (!Directed)
        {
            AdjacencyList[vertex2].Add(vertex1);
        }
    }
}


class Result
{
    public static long MinimumCost(Graph graph, long costLibrary, long costRoad)
    {
        long totalCost = 0;

         return totalCost;
    }

    private static int DepthFirstSearchUtil(Graph graph, bool[] visited, int currentVertex)
    {
        int count = 1;
 
        return count;
    }

    //public static long roadsAndLibraries(int n, int c_lib, int c_road, List<List<int>> cities)
    public static long roadsAndLibraries(int numCities, int costLibrary, int costRoad, List<List<int>> cities)
    {
        var graph = new Graph(numCities, false);

        foreach (var city in cities)
        {
            graph.AddEdge(city[0] - 1, city[1] - 1);
        }

        long minimumCost = MinimumCost(graph, costLibrary, costRoad);

        return minimumCost;
    }
}


class Solution
{
    public static void Main(string[] args)
    {
        TextWriter textWriter = new StreamWriter(@System.Environment.GetEnvironmentVariable("OUTPUT_PATH"), true);

        int q = Convert.ToInt32(Console.ReadLine().Trim());

        for (int qItr = 0; qItr < q; qItr++)
        {
            string[] firstMultipleInput = Console.ReadLine().TrimEnd().Split(' ');

            int n = Convert.ToInt32(firstMultipleInput[0]);

            int m = Convert.ToInt32(firstMultipleInput[1]);

            int c_lib = Convert.ToInt32(firstMultipleInput[2]);

            int c_road = Convert.ToInt32(firstMultipleInput[3]);

            List<List<int>> cities = new List<List<int>>();

            for (int i = 0; i < m; i++)
            {
                cities.Add(Console.ReadLine().TrimEnd().Split(' ').ToList().Select(citiesTemp => Convert.ToInt32(citiesTemp)).ToList());
            }

            long result = Result.roadsAndLibraries(n, c_lib, c_road, cities);

            textWriter.WriteLine(result);
        }

        textWriter.Flush();
        textWriter.Close();
    }
}

File: Roads-and-Libraries.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;


class Graph {
    boolean directed;
    int numVertices;
    ArrayList<LinkedList<Integer>> adjacencyList;

    public Graph(int numVertices, boolean directed) {
        this.numVertices = numVertices;
        this.directed = directed;
        adjacencyList = new ArrayList<>();
        for (int i = 0; i < numVertices; i++) {
            adjacencyList.add(new LinkedList<>());
        }
    }

    public void addEdge(int vertex1, int vertex2) {
        adjacencyList.get(vertex1).add(vertex2);
        if (!directed) {
            adjacencyList.get(vertex2).add(vertex1);
        }
    }
}


class Result {
    public static int depthFirstSearchUtil(Graph graph, boolean[] visited, int currentVertex) {
        int count = 1;
        visited[currentVertex] = true;

        for (long neighbor : graph.adjacencyList.get(currentVertex)) {
            if (!visited[(int) neighbor]) {
                count += depthFirstSearchUtil(graph, visited, (int) neighbor);
            }
        }

        return count;
    }

    public static long minimumCost(Graph graph, long costLibrary, long costRoad) {
        if (costLibrary < costRoad) {
            return costLibrary * graph.numVertices;
        }

        boolean[] visited = new boolean[(int) graph.numVertices];
        long totalCost = 0;

        for (int i = 0; i < graph.numVertices; ++i) {
            if (!visited[i]) {
                int numConnectedVertices = depthFirstSearchUtil(graph, visited, i);
                totalCost += (numConnectedVertices - 1) * costRoad;
                totalCost += costLibrary;
            }
        }

        return totalCost;
    }
    
    // public static long roadsAndLibraries(int n, int c_lib, int c_road, List<List<Integer>> cities) {
    public static long roadsAndLibraries(int numCities, int costLibrary, int costRoad, List<List<Integer>> cities) {
        Graph graph = new Graph(numCities, false);

        for (List<Integer> city : cities) {
            graph.addEdge(city.get(0) - 1, city.get(1) - 1);
        }

        return minimumCost(graph, costLibrary, costRoad);
    }
}


public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int q = Integer.parseInt(bufferedReader.readLine().trim());

        IntStream.range(0, q).forEach(qItr -> {
            try {
                String[] firstMultipleInput = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");

                int n = Integer.parseInt(firstMultipleInput[0]);

                int m = Integer.parseInt(firstMultipleInput[1]);

                int c_lib = Integer.parseInt(firstMultipleInput[2]);

                int c_road = Integer.parseInt(firstMultipleInput[3]);

                List<List<Integer>> cities = new ArrayList<>();

                IntStream.range(0, m).forEach(i -> {
                    try {
                        cities.add(
                            Stream.of(bufferedReader.readLine().replaceAll("\\s+$", "").split(" "))
                                .map(Integer::parseInt)
                                .collect(toList())
                        );
                    } catch (IOException ex) {
                        throw new RuntimeException(ex);
                    }
                });

                long result = Result.roadsAndLibraries(n, c_lib, c_road, cities);

                bufferedWriter.write(String.valueOf(result));
                bufferedWriter.newLine();
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        });

        bufferedReader.close();
        bufferedWriter.close();
    }
}
